### PR TITLE
Fix #3293: Introduced Developer Options Starter

### LINKED ISSUE #3293: Introduce Developer Options Starter
Introduce `DeveloperOptionsStarter` module. It will be responsible for checking for availability of **DevMode**.

**Target PR date:** 2021-06-09
**Target completion date:** 2021-06-12

### PR DESCRIPTION
**Pull request by yashraj-01**
## Explanation

Fixes #3293: Introduced Developer Options Starter. This PR also introduces the initial label test for `DeveloperOptionsActivity`.

**Screenshots**

- Espresso tests passing for `DeveloperOptionsActivityTest`
Screenshot from 2021-06-22 20-26-22

- Robolectric tests passing for `DeveloperOptionsActivityTest`
Screenshot from 2021-06-22 20-27-45

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt]: Nit: Can be renamed to something like `DebugApplicationComponent` since the mode name will be debug, not dev.

File: app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt]: Nit: in debug mode

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: nit: `debugComponent` sounds better.

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: nit: `enableDebugMode`

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (Sarthak2601) [inline app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: It might be worth mentioning what we're doing here in a comment.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Nice work @yashraj-01!. Suggested some nit changes -- PTAL.

File: app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt]: Done.

File: app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/application/ApplicationComponentForDevMode.kt]: Done.

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: Now outdated.

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: Done.

File: app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/application/OppiaApplication.kt]: Now outdated.

**Comment by yashraj-01** [comment]
I have pushed the latest changes. PTAL @Sarthak2601. Thanks.

**Review by Sarthak2601 — APPROVED** [review]
LGTM. 
Note: Having multiple application components based on build type depends on @BenHenning's approval.

**Review by Sarthak2601 — CHANGES_REQUESTED** [review]
Blocking merge until approach approval.

**Comment by yashraj-01** [comment]
@BenHenning I have updated the approach according to your review on the proposal and your suggestions on the chat. PTAL. Thanks.

**Comment by yashraj-01** [comment]
@rt4914 can you please provide your code owner review? Thanks.

**Comment by rt4914** [comment]
> @rt4914 can you please provide your code owner review? Thanks.

@yashraj-01 Please do keep the `Assignees` section up-to-date. https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section

File: app/src/main/AndroidManifest.xml
Comment (rt4914) [inline app/src/main/AndroidManifest.xml:200]: Please add a test case for this label similar to `testFaqListActivity_hasCorrectActivityLabel`

File: app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarterImpl.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarterImpl.kt]: Add Kdoc for this.

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt:11]: @Sarthak2601 Was this suppose to be a separate module or just separate package?

File: app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt]: Why do we need this? Why can't we just call `DeveloperOptionsActivity.createDeveloperOptionsActivityIntent(context, internalProfileId)` directly similar to our current code base ?

**Review by rt4914 — CHANGES_REQUESTED** [review]
@yashraj-01

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt
Comment (Sarthak2601) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt:11]: Earlier it was supposed to be a separate module but now we've shifted it to separate package approach.

File: app/src/main/AndroidManifest.xml
Comment (yashraj-01) [reply app/src/main/AndroidManifest.xml:200]: Done. PTAL.

File: app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarterImpl.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarterImpl.kt]: Done. PTAL.

File: app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt
Comment (yashraj-01) [reply app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt]: It is necessary to follow this approach as for non-debug builds, we won't be including DeveloperOptionsActivity and all the related files to it in our build. So, if we follow the conventional approach the app will throw a build time error. Having this Starter provides Optional binding for the intent which means if the Dev modules are not found in the build it won't throw an error. It will just treat it as if it is not present. Hope this clarifies the approach. Thanks.

**Comment by yashraj-01** [comment]
@rt4914 I have addressed all your comments. PTAL. Thanks.

**Comment by seanlip** [comment]
Hi folks -- is this still waiting on @BenHenning or just @rt4914?

Also @rt4914 could you PTAL? This seems to block 4 other PRs from what I can tell -- https://github.com/oppia/oppia-android/pulls/yashraj-01

Thanks!

File: app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/devoptions/DeveloperOptionsActivity.kt:11]: Okay.

File: app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/drawer/DeveloperOptionsStarter.kt]: Makes sense. Thanks.

**Review by rt4914 — APPROVED** [review]
LGTM, thanks.

**Comment by Sarthak2601** [comment]
@yashraj-01 shouldn't we have all files related to Dev Options Menu in the `devOptions` package? Seems to me that some files are in the drawer package.

**Comment by yashraj-01** [comment]
> @yashraj-01 shouldn't we have all files related to Dev Options Menu in the `devOptions` package? Seems to me that some files are in the drawer package.

Moved them to `devoptions` package. PTAL.

**Review by Sarthak2601 — APPROVED** [review]
LGTM.

**Pull request merged**
